### PR TITLE
bug: fix TypeMismatch panic at compaction

### LINKED ISSUE #696: bug: compaction panic
When working on #645, I found this bug where the compactor panics after two insertions.

image

I have looked into this. Will provide a more detailed description of the bug, and submit a PR to fix it.

### PR DESCRIPTION
**Pull request by yinfredyue**
Trying to fix #696

### CODE REVIEW

**Comment by skyzh** [comment]
`From<&array::ArrayImpl>` is implemented on concrete types (I32Array, Utf8Array, etc.) instead of the generic `Array` trait. Therefore, you have to use macros to expand them instead of calling it on traits.

**Comment by yinfredyue** [comment]
PTAL @skyzh

**Comment by yinfredyue** [comment]
SQL execution before this fix:

image

SQL execution after the fix: 

image

File: src/array/shuffle_ext.rs [hunk 1]
```diff
@@ -141,11 +133,10 @@ macro_rules! impl_array_impl_shuffle_ext {
                 match self {
                     $(
                         Self::$Abc(builder) => {
-                            let typed_arrays = arrays
-                                .iter()
-                                .map(|x| x.try_into().unwrap())
-                                .collect::<SmallVec<[_; 8]>>();
-                            builder.pick_from_multiple(&typed_arrays, logical_rows);
+                            for (idx, row) in logical_rows {
+                                let arr: & $AbcArray = (&arrays[*idx]).try_into().unwrap();
```
Comment (skyzh) [inline src/array/shuffle_ext.rs:137]: This is inefficient because we need to branching inside a loop to do the cast every time. Can we move arr outside?

**Pull request closed without merging**